### PR TITLE
`Iris`: Integrate with Hestia to generate descriptions for code hints

### PR DESCRIPTION
**Pull request by Hialus**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

I had the idea to integrate Iris with Hestia a while ago. Iris is now in a state where a first version of this is actually possible.
The goal is to allow editors to automatically generate descriptions for code hints, so that the code hint createion becomes even less work.

### Description

To do this I had to implement a couple things:
- Iris now supports multiple session types
- The old sessions are now chat sessions
- The new sessions for hestia are hestia sessions
- The session service is now supported by two sub services for the respective session types
- The new hestia session service manages the generation of the descriptions

### Steps for Testing

Prerequisites:
- 1 Instructor/Editor
- 1 Java Programming Exercise with Code Coverage enabled
- Only possible on TS9

1. Log in to Artemis
2. Navigate to the programming exercise instructor page
3. Wait for the build plans for solution and template to run through
4. Go to "Hints"
5. Go to "Generate Code Hints"
6. Create the solution code snippets
7. Creat the code hints
8. Edit one of the code hints
9. Press the new button to generate the descriptions
10. Wait a bit
11. The description and content fields should now be filled
12. Save
13. The detail page should show them now
14. If you want view the hint as a student

#### Exam Mode Testing

Hints are not supported for the exam mode

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| IrisChatSessionService.java | 95% | ✅                           |
| IrisHestiaSessionService.java | 93% | ✅              |
| IrisSessionService.java | 86% | ✅              |
| IrisWebsocketService.java | 88% | ✅              |
| IrisMessageService.java | 100% | ✅              |
| CodeHintService.java | 96% | ✅              |
| exercise-hint-update.component.ts | 84% | ✅              |

### Screenshots

The new button: 
grafik

Screencast: 
https://github.com/ls1intum/Artemis/assets/18218285/293f2279-a4a5-4cfb-aede-de70276f5862

### CODE REVIEW

**Review by RY997 — DISMISSED** [review]
Tested on ts9, but when click "generate description & content" button I got this error
image

**Comment by Hialus** [comment]
> Tested on ts9, but when click "generate description & content" button I got this error image

Sorry. I forgot to update the test server when the profile name changed. Should be fixed now.

File: src/main/java/de/tum/in/www1/artemis/domain/iris/IrisChatSession.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return user;
+    }
+
+    public void setUser(User user) {
+        this.user = user;
+    }
+
+    @Override
+    public String toString() {
+        return "IrisChatSession{" + "id=" + getId() + "exercise=" + (exercise == null ? "null" : exercise.getId()) + ", user=" + (user == null ? "null" : user.getName()) + '}';
```
Comment (SolizerCodes) [inline src/main/java/de/tum/in/www1/artemis/domain/iris/IrisChatSession.java:46]:
```suggestion
        return "IrisChatSession{" + "id=" + getId() + ", exercise=" + (exercise == null ? "null" : exercise.getId()) + ", user=" + (user == null ? "null" : user.getName()) + '}';
```
, is missing here

File: src/main/java/de/tum/in/www1/artemis/domain/iris/IrisHestiaSession.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return codeHint;
+    }
+
+    public void setCodeHint(CodeHint codeHint) {
+        this.codeHint = codeHint;
+    }
+
+    @Override
+    public String toString() {
+        return "IrisHestiaSession{" + "id=" + getId() + "codeHint=" + (codeHint == null ? "null" : codeHint.getId()) + '}';
```
Comment (SolizerCodes) [inline src/main/java/de/tum/in/www1/artemis/domain/iris/IrisHestiaSession.java:35]:
```suggestion
        return "IrisHestiaSession{" + "id=" + getId() + ", codeHint=" + (codeHint == null ? "null" : codeHint.getId()) + '}';
```

File: src/main/resources/config/liquibase/master.xml [hunk 3]
```diff
@@ -32,6 +32,7 @@
     <include file="classpath:config/liquibase/changelog/20230430235959_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20230508114600_changelog.xml" relativeToChangelogFile="false"/>
     <include file="classpath:config/liquibase/changelog/20230517171700_changelog.xml" relativeToChangelogFile="false"/>
+    <include file="classpath:config/liquibase/changelog/20230527010101_changelog.xml" relativeToChangelogFile="false"/>
```
Comment (SolizerCodes) [inline src/main/resources/config/liquibase/master.xml:35]: Not sure but I think they should be ordered according to datetime ascending

**Review by SolizerCodes — COMMENTED** [review]
Found three small things in the code

**Review by RY997 — COMMENTED** [review]
Now the generate button works, and it's very cool. But I did't find the hints in online & offline editor which I assumed the hints should be shown in the code file when students clone the repo or I'm wrong.
屏幕截图 2023-05-29 220211
image

**Comment by Hialus** [comment]
> Now the generate button works, and it's very cool. But I did't find the hints in online & offline editor which I assumed the hints should be shown in the code file when students clone the repo or I'm wrong. 屏幕截图 2023-05-29 220211 image

Hints are actually displayed on Artemis in the problem statement view. Here's how you can view them:
- There is a display threshold and a task association for each hint
- If the student fails the task x times the hint will become available over a button in the bottom right corner of the problem statement
    - This only happens if all tasks before that specific task are successfull.
- Once you click on the button you can activate the hint and see its content
- If you set the threshold to 0, the hint will always be visible as long as you submitted the exercise at least once and it is not a build failure

Though keep in mind that I did not change this, so you don't have to test this :)

File: src/main/java/de/tum/in/www1/artemis/domain/iris/IrisChatSession.java (hunk 1 above)
Comment (Hialus) [reply src/main/java/de/tum/in/www1/artemis/domain/iris/IrisChatSession.java:46]: Good find!

**Review by RY997 — APPROVED** [review]
Thanks for the detailed explanation! That's a really cool feature.

**Review by SabrinaGlatz — APPROVED** [review]
Reviewed the code and tested it on ts9. It works as expected

**Review by SolizerCodes — APPROVED** [review]
Code LGTM

**Pull request merged**